### PR TITLE
QBFT - Create extradata with round information only

### PR DESCRIPTION
**Pull request by usmansaleem**
## PR description
Create extra data with round information only when using a contract to obtain vote and validators information.
 
## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java [hunk 1]
```diff
@@ -45,27 +54,45 @@ public boolean validate(
       final BftContext bftContext = context.getConsensusState(BftContext.class);
       final BftExtraData bftExtraData = bftContext.getBlockInterface().getExtraData(header);
 
-      final NavigableSet<Address> sortedReportedValidators =
-          new TreeSet<>(bftExtraData.getValidators());
+      // validators and vote may be empty if they are supplied by a contract.
```
Comment (jframe) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java:57]: This comment is slightly misleading, the validators and vote must be empty.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java [hunk 2]
```diff
@@ -37,6 +37,15 @@
 public class BftValidatorsValidationRule implements AttachedBlockHeaderValidationRule {
 
   private static final Logger LOGGER = LogManager.getLogger();
+  private final boolean extraDataValidatorsAndVoteMustBeEmpty;
+
+  public BftValidatorsValidationRule() {
+    this(false);
+  }
+
+  public BftValidatorsValidationRule(final boolean extraDataValidatorsAndVoteMustBeEmpty) {
```
Comment (jframe) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java:46]: This rule is quite specific to qbft and so I don't think it should be part of the common rules. Would rather the common bft code didn't know anything about contract based voting/validator governance.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/BftBlockCreatorFactory.java [hunk 3]
```diff
@@ -102,22 +105,35 @@ public Wei getMinTransactionGasPrice() {
...
-
-    final BftExtraData extraData =
-        new BftExtraData(
-            ConsensusHelpers.zeroLeftPad(vanityData, BftExtraDataCodec.EXTRA_VANITY_LENGTH),
-            Collections.emptyList(),
-            toVote(proposal),
-            round,
-            validators);
+    final BftExtraData extraData;
+    if (extraDataWithRoundInformationOnly) {
```
Comment (jframe) [inline consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/BftBlockCreatorFactory.java:109]: Would prefer this wasn't part of the base BftBlockCreatorFactory and we had a specific QbftBlockCreatorFactory as it's specific to Qbft.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java (hunk 1 above)
Comment (usmansaleem) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java]: updated.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java (hunk 2 above)
Comment (usmansaleem) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/headervalidationrules/BftValidatorsValidationRule.java]: refactored.

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/BftBlockCreatorFactory.java (hunk 3 above)
Comment (usmansaleem) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/BftBlockCreatorFactory.java]: refactored.

File: consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/QbftBlockHeaderValidationRulesetFactoryTest.java [hunk 4]
```diff
@@ -63,7 +63,7 @@ public void bftValidateHeaderPasses() {
         getPresetHeaderBuilder(2, proposerNodeKey, validators, parentHeader).buildHeader();
 
     final BlockHeaderValidator validator =
-        QbftBlockHeaderValidationRulesetFactory.blockHeaderValidator(5).build();
+        new QbftBlockHeaderValidationRulesetFactory(false).blockHeaderValidator(5).build();
```
Comment (jframe) [inline consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/QbftBlockHeaderValidationRulesetFactoryTest.java:66]: Can some tests be added for when we are using contract based validation rules

Comment (usmansaleem) [reply consensus/qbft/src/test/java/org/hyperledger/besu/consensus/qbft/QbftBlockHeaderValidationRulesetFactoryTest.java:66]: added couple of new test classes to cover new logic.

**Review by jframe — APPROVED** [review]
LGTM

**Pull request merged**